src/crimson/osd/main.cc

```diff
@@ -60,45 +61,61 @@ int main(int argc, char* argv[])
     usage(argv[0]);
     return EXIT_SUCCESS;
   }
-  std::string cluster;
+  std::string cluster_name;
   std::string conf_file_list;
   // ceph_argparse_early_args() could _exit(), while local_conf() won't ready
   // until it's started. so do the boilerplate-settings parsing here.
   auto init_params = ceph_argparse_early_args(ceph_args,
                                               CEPH_ENTITY_TYPE_OSD,
-                                              &cluster,
+                                              &cluster_name,
                                               &conf_file_list);
   seastar::sharded<OSD> osd;
+  seastar::sharded<ceph::net::SocketMessenger> cluster_msgr, client_msgr;
+  seastar::sharded<ceph::net::SocketMessenger> hb_front_msgr, hb_back_msgr;
   using ceph::common::sharded_conf;
   using ceph::common::sharded_perf_coll;
   using ceph::common::local_conf;
   try {
     return app.run_deprecated(app_args.size(), const_cast<char**>(app_args.data()), [&] {
       auto& config = app.configuration();
-      seastar::engine().at_exit([] {
-        return sharded_conf().stop();
-      });
-      seastar::engine().at_exit([] {
-        return sharded_perf_coll().stop();
-      });
-      seastar::engine().at_exit([&] {
-       return osd.stop();
-      });
-      return sharded_conf().start(init_params.name, cluster).then([] {
-        return sharded_perf_coll().start();
-      }).then([&conf_file_list] {
-        return local_conf().parse_config_files(conf_file_list);
-      }).then([&] {
-        return osd.start_single(std::stoi(local_conf()->name.get_id()),
-                                static_cast<uint32_t>(getpid()));
-      }).then([&osd, mkfs = config.count("mkfs")] {
-        if (mkfs) {
-          return osd.invoke_on(0, &OSD::mkfs,
+      return seastar::async([&] {
+        sharded_conf().start(init_params.name, cluster_name).get();
+        sharded_perf_coll().start().get();
+        local_conf().parse_config_files(conf_file_list).get();
+        const int whoami = std::stoi(local_conf()->name.get_id());
+        const auto nonce = static_cast<uint32_t>(getpid());
+        const auto shard = seastar::engine().cpu_id();
+        cluster_msgr.start(entity_name_t::OSD(whoami), string{"cluster"}, nonce, shard).get();
```

Comment:
@cyx1231st finally, i realized that we should create messengers in `main.cc`. you were right.